cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java

```diff
@@ -0,0 +1,194 @@
+          // check for program Id for the run record, if null then it is invalid program type.
+          ProgramId targetProgramId = programLifecycleService.retrieveProgramIdForRunRecord(programType, runId);
+
+          // Check if run id is for the right program type
+          if (targetProgramId != null) {
+            runIdToProgramId.put(runId, targetProgramId);
+            return true;
+          } else {
+            return false;
+          }
+        }
+      });
+
+    // don't correct run records for programs running inside a workflow
+    // for instance, a MapReduce running in a Workflow will not be contained in the runtime info in this class
+    invalidRunRecords = Collections2.filter(invalidRunRecords, new com.google.common.base.Predicate<RunRecordMeta>() {
+      @Override
+      public boolean apply(RunRecordMeta invalidRunRecordMeta) {
+        boolean shouldCorrect = programLifecycleService.shouldCorrectForWorkflowChildren(invalidRunRecordMeta,
+                                                                                         processedInvalidRunRecordIds);
+        if (!shouldCorrect) {
+          LOG.trace("Will not correct invalid run record {} since it's parent workflow still running.",
+                    invalidRunRecordMeta);
+          return false;
+        }
+        return true;
+      }
+    });
+
+    LOG.trace("End getting invalid run records.");
+
+    if (!invalidRunRecords.isEmpty()) {
+      LOG.warn("Found {} RunRecords with RUNNING status and the program not actually running for program type {}",
+               invalidRunRecords.size(), programType.getPrettyName());
+    } else {
+      LOG.trace("No RunRecords found with RUNNING status and the program not actually running for program type {}",
+                programType.getPrettyName());
+    }
+
+    // Now lets correct the invalid RunRecords
+    for (RunRecordMeta invalidRunRecordMeta : invalidRunRecords) {
+      String runId = invalidRunRecordMeta.getPid();
+      ProgramId targetProgramId = runIdToProgramId.get(runId);
+
+      boolean updated = store.compareAndSetStatus(targetProgramId, runId, ProgramController.State.ALIVE.getRunStatus(),
+                                                  ProgramController.State.ERROR.getRunStatus());
+      if (updated) {
+        LOG.warn("Fixed RunRecord {} for program {} with RUNNING status because the program was not " +
+                   "actually running",
+                 runId, targetProgramId);
+
+        processedInvalidRunRecordIds.add(runId);
+      }
+    }
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting RunRecordCorrectorService");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Stopping RunRecordCorrectorService");
+  }
+
+  /**
+   * Helper class to run in separate thread to validate the invalid running run records
+   */
+  class RunRecordsCorrectorRunnable implements Runnable {
```

Comment:
This runnable should only be in the distributed implementation, since only in distributed mode you need to run it periodically, right? The `validateAndCorrectRunningRunRecords` should be a protected method for sub-class to use, such that in local mode, it can be called directly upon start, while in distributed mode, it gets called periodically.